Seata error when executing select for update statement with mybatis

- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
Seata error when executing select for update statement with mybatis. If I remove the for update, the service will not report an error.

### Ⅱ. Describe what happened

  If there is an exception, please attach the exception trace:

```
### Error querying database.  Cause: java.sql.SQLException: java.lang.ClassCastException: class java.lang.Boolean cannot be cast to class java.sql.ResultSet (java.lang.Boolean is in module java.base of loader 'bootstrap'; java.sql.ResultSet is in module java.sql of loader 'platform')
### The error may exist in file [D:\workspace\LearnWorkspace\fescar-samples\dubbo\target\classes\mapper\AccountMapper.xml]
### The error may involve com.alibaba.fescar.samples.dubbo.dao.AccountDao.getAccountByUserIdForLock-Inline
### The error occurred while setting parameters
### SQL: select * from account_tbl where user_id = ? for update
### Cause: java.sql.SQLException: java.lang.ClassCastException: class java.lang.Boolean cannot be cast to class java.sql.ResultSet (java.lang.Boolean is in module java.base of loader 'bootstrap'; java.sql.ResultSet is in module java.sql of loader 'platform')
; uncategorized SQLException; SQL state [null]; error code [0]; java.lang.ClassCastException: class java.lang.Boolean cannot be cast to class java.sql.ResultSet (java.lang.Boolean is in module java.base of loader 'bootstrap'; java.sql.ResultSet is in module java.sql of loader 'platform'); nested exception is java.sql.SQLException: java.lang.ClassCastException: class java.lang.Boolean cannot be cast to class java.sql.ResultSet (java.lang.Boolean is in module java.base of loader 'bootstrap'; java.sql.ResultSet is in module java.sql of loader 'platform')
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:89)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:81)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:81)
	at org.mybatis.spring.MyBatisExceptionTranslator.translateExceptionIfPossible(MyBatisExceptionTranslator.java:73)
	at org.mybatis.spring.SqlSessionTemplate$SqlSessionInterceptor.invoke(SqlSessionTemplate.java:446)
	at com.sun.proxy.$Proxy13.selectOne(Unknown Source)
	at org.mybatis.spring.SqlSessionTemplate.selectOne(SqlSessionTemplate.java:166)
	at org.apache.ibatis.binding.MapperMethod.execute(MapperMethod.java:87)
	at org.apache.ibatis.binding.MapperProxy.invoke(MapperProxy.java:58)
	at com.sun.proxy.$Proxy14.getAccountByUserIdForLock(Unknown Source)
	at com.alibaba.fescar.samples.dubbo.service.impl.AccountServiceImpl$2.doInTransaction(AccountServiceImpl.java:102)
	... 33 more
Caused by: java.sql.SQLException: java.lang.ClassCastException: class java.lang.Boolean cannot be cast to class java.sql.ResultSet (java.lang.Boolean is in module java.base of loader 'bootstrap'; java.sql.ResultSet is in module java.sql of loader 'platform')
	at io.seata.rm.datasource.exec.ExecuteTemplate.execute(ExecuteTemplate.java:107)
	at io.seata.rm.datasource.exec.ExecuteTemplate.execute(ExecuteTemplate.java:48)
	at io.seata.rm.datasource.PreparedStatementProxy.execute(PreparedStatementProxy.java:66)
	at org.apache.ibatis.executor.statement.PreparedStatementHandler.query(PreparedStatementHandler.java:64)
	at org.apache.ibatis.executor.statement.RoutingStatementHandler.query(RoutingStatementHandler.java:79)
	at org.apache.ibatis.executor.SimpleExecutor.doQuery(SimpleExecutor.java:63)
	at org.apache.ibatis.executor.BaseExecutor.queryFromDatabase(BaseExecutor.java:324)
	at org.apache.ibatis.executor.BaseExecutor.query(BaseExecutor.java:156)
	at org.apache.ibatis.executor.CachingExecutor.query(CachingExecutor.java:109)
	at org.apache.ibatis.executor.CachingExecutor.query(CachingExecutor.java:83)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:147)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:140)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectOne(DefaultSqlSession.java:76)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.mybatis.spring.SqlSessionTemplate$SqlSessionInterceptor.invoke(SqlSessionTemplate.java:433)
	... 39 more
Caused by: java.lang.ClassCastException: class java.lang.Boolean cannot be cast to class java.sql.ResultSet (java.lang.Boolean is in module java.base of loader 'bootstrap'; java.sql.ResultSet is in module java.sql of loader 'platform')
	at io.seata.rm.datasource.exec.SelectForUpdateExecutor.doExecute(SelectForUpdateExecutor.java:86)
	at io.seata.rm.datasource.exec.BaseTransactionalExecutor.execute(BaseTransactionalExecutor.java:87)
	at io.seata.rm.datasource.exec.ExecuteTemplate.execute(ExecuteTemplate.java:102)
	... 56 more

```

--- a/rm-datasource/src/main/java/io/seata/rm/datasource/exec/ExecuteTemplate.java
+++ b/rm-datasource/src/main/java/io/seata/rm/datasource/exec/ExecuteTemplate.java
@@ -90,7 +90,7 @@ public class ExecuteTemplate {
                     executor = new DeleteExecutor<T, S>(statementProxy, statementCallback, sqlRecognizer);
                     break;
                 case SELECT_FOR_UPDATE:
-                    executor = new SelectForUpdateExecutor(statementProxy, statementCallback, sqlRecognizer);
+                    executor = new SelectForUpdateExecutor<T, S>(statementProxy, statementCallback, sqlRecognizer);
                     break;
                 default:
                     executor = new PlainExecutor<T, S>(statementProxy, statementCallback);
@@ -100,7 +100,6 @@ public class ExecuteTemplate {
         T rs = null;
         try {
             rs = executor.execute(args);
-
         } catch (Throwable ex) {
             if (!(ex instanceof SQLException)) {
                 // Turn other exception into SQLException

--- a/rm-datasource/src/main/java/io/seata/rm/datasource/exec/SelectForUpdateExecutor.java
+++ b/rm-datasource/src/main/java/io/seata/rm/datasource/exec/SelectForUpdateExecutor.java
@@ -38,7 +38,7 @@ import io.seata.rm.datasource.sql.struct.TableRecords;
  *
  * @param <S> the type parameter
  */
-public class SelectForUpdateExecutor<S extends Statement> extends BaseTransactionalExecutor<ResultSet, S> {
+public class SelectForUpdateExecutor<T, S extends Statement> extends BaseTransactionalExecutor<T, S> {
 
     /**
      * Instantiates a new Select for update executor.
@@ -47,17 +47,17 @@ public class SelectForUpdateExecutor<S extends Statement> extends BaseTransactio
      * @param statementCallback the statement callback
      * @param sqlRecognizer     the sql recognizer
      */
-    public SelectForUpdateExecutor(StatementProxy<S> statementProxy, StatementCallback<ResultSet, S> statementCallback,
+    public SelectForUpdateExecutor(StatementProxy<S> statementProxy, StatementCallback<T, S> statementCallback,
                                    SQLRecognizer sqlRecognizer) {
         super(statementProxy, statementCallback, sqlRecognizer);
     }
 
     @Override
-    public Object doExecute(Object... args) throws Throwable {
+    public T doExecute(Object... args) throws Throwable {
         SQLSelectRecognizer recognizer = (SQLSelectRecognizer)sqlRecognizer;
 
         Connection conn = statementProxy.getConnection();
-        ResultSet rs = null;
+        T rs = null;
         Savepoint sp = null;
         LockRetryController lockRetryController = new LockRetryController();
         boolean originalAutoCommit = conn.getAutoCommit();
@@ -83,6 +83,9 @@ public class SelectForUpdateExecutor<S extends Statement> extends BaseTransactio
                 conn.setAutoCommit(false);
             }
             sp = conn.setSavepoint();
+            // #870
+            // execute return Boolean
+            // executeQuery return ResultSet
             rs = statementCallback.execute(statementProxy.getTargetStatement(), args);
 
             while (true) {

